### PR TITLE
8336406: Add support for FP16 binary operations

### PR DESCRIPTION
**Pull request by Bhavana-Kilambi**
This patch adds middle end support in C2 for a few FP16 binary operations, namely - subtract, multiply, divide, min and max. It also adds aarch64 backend support for these operations.

Tested JTREG tests -
hotspot/jtreg/compiler/vectorization/TestFloat16VectorOps.java, hotspot/jtreg/compiler/vectorization/TestFloat16VectorReinterpretConv.java, hotspot/jtreg/compiler/intrinsics/float16 and test/jdk/java/lang/Float16 and they successfully pass on aarch64 and x86 machines.

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Issue
 * [JDK-8336406](https://bugs.openjdk.org/browse/JDK-8336406): Add support for FP16 binary operations (**Enhancement** - P4)

### Reviewers
 * [Jatin Bhateja](https://openjdk.org/census#jbhateja) (@jatin-bhateja - Committer) ⚠️ Review applies to [914dfc24](https://git.openjdk.org/valhalla/pull/1175/files/914dfc2477ae3a4119a6f7a749f2b7b2f4840963)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/valhalla.git pull/1175/head:pull/1175` \
`$ git checkout pull/1175`

Update a local copy of the PR: \
`$ git checkout pull/1175` \
`$ git pull https://git.openjdk.org/valhalla.git pull/1175/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1175`

View PR using the GUI difftool: \
`$ git pr show -t 1175`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/valhalla/pull/1175.diff">https://git.openjdk.org/valhalla/pull/1175.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/valhalla/pull/1175#issuecomment-2245302076)

### CODE REVIEW

**Comment by Bhavana-Kilambi** [comment]
Hi @jatin-bhateja , can I please get some review for this patch? Thank you !

**Comment by jatin-bhateja** [comment]
Hi @Bhavana-Kilambi ,

Your patch looks good to me, just one comment.
Once it gets integrated I will add x86 backend support for newly supported intrinsic.
Best Regards,
Jatin

File: src/hotspot/share/opto/mulnode.hpp [hunk 1]
```diff
@@ -143,6 +143,18 @@ class MulFNode : public MulNode {
   virtual uint ideal_reg() const { return Op_RegF; }
 };
 
+//------------------------------MulHFNode---------------------------------------
+// Multiply 2 half floats
+class MulHFNode : public MulFNode {
+public:
```
[the remark this answers is not in the document]
Comment (Bhavana-Kilambi) [reply src/hotspot/share/opto/mulnode.hpp:149]: Thank you for the review. I agree with your point and somehow missed this. I will make changes and upload a new PS soon.

[the remark this answers is not in the document]
Comment (Bhavana-Kilambi) [reply src/hotspot/share/opto/mulnode.hpp:149]: Hi @jatin-bhateja , I have made the changes to MulHF::Ideal() as suggested. This avoids generation of pure Floating point IR during Ideal optimizations of MulHF. I have checked other binary operations as well to make sure we are not generating any pure floating point IR during optimization. I can see one in DivFNode::Ideal() - https://github.com/openjdk/valhalla/blob/e45febcdb3f6bd8ca139f6c12126d52968c7c2f9/src/hotspot/share/opto/divnode.cpp#L749 which generates MulFNode when dividing with 2. I tried this specific pattern for half float and it generates correct answer but performs the entire optimization in float IR (the input FP16 value is converted to FP32). For now, I don't think we can change this as the constant value 2 is interpreted as a TypeF floating point value and we do not have the corresponding type for Float16 at the moment. The resultant output is of course correct and thus I have not made any changes to it. 
Other binary operations seem to be fine to be performed on Float16 values as well.

**Review by jatin-bhateja — APPROVED** [review]
Please address the closing comments and integrate.

Patch looks good to me. Will add x86 backend support for newly added intrinsic.

**Comment by jatin-bhateja** [comment]
Please address the closing comments and integrate.

Patch looks good to me. Will add x86 backend support for newly added intrinsic.

**Comment by Bhavana-Kilambi** [comment]
Thank you for the review. 
/integrate

**Comment by Bhavana-Kilambi** [comment]
Hi @jatin-bhateja, I just noticed that the IR tests for x *2 -> x + x already exist for float and double here - `test/hotspot/jtreg/compiler/c2/irTests/TestMulNodeIdealization.java`. Shall I update this test with the Half float Ideal tests and remove this file introduced in this patch - `test/hotspot/jtreg/compiler/c2/irTests/MulHFNodeIdealizationTests.java`?
I am thinking, we will have to pass `--enable-preview` flag for the whole file which is applicable even for float and double tests which may not need it and we might not be able to test these testcases in older JDKs. What do you suggest? Is it better to keep this test separate for half-float or integrate it with float and double tests?

**Pull request closed without merging**